config/initializers/validators/certificate_validator.rb

```diff
@@ -23,19 +23,31 @@ def certificate_is_valid
     certificate_key_is_supported
   end
 
-  def factory
-    Utilities::Certificate::CertificateFactory
+  def convert_value_to_x509_certificate
+    begin
+      OpenSSL::X509::Certificate.new(value)
+    rescue
+      OpenSSL::X509::Certificate.new(Base64.decode64(value))
+    end
   end
 
   def x509_certificate
-    x509 = factory.x509_certificate(value)
+    if value != @last_converted_value || @x509.blank?
+      @x509 = convert_value_to_x509_certificate
+      @last_converted_value = value
+    end
+    @x509
   rescue
-    errors.add(:certificate, 'is not a valid x509 certificate') if x509.nil?
-    x509
+    errors.add(:certificate, 'is not a valid x509 certificate') if @x509.nil?
+    @x509
```

Comment:
I don't there is a point in returning the certificate if it's nil. Return nil or just nothing

Reply:
Well am just returning the **variables' value** which is equal to **nil** could have written nil instead of @x509 which I suppose is more explicit am not sure just
`errors.add(:certificate, 'is not a valid x509 certificate')` is nil, I think I have to put something after

@x509 = nil
yeah I could have put **nil** which is more obvious, I was just trying to following ruby's philosophy
not using return unless it is necessary hence that why I didn't put return @x509e
I'll put nil to be explicit

Reply:
the variable's value is very likely nil or errored out, so there is no point returning it. If you don't return anything, that will be nil automatically - we don't have to be explicit

Reply:
I can confirm this is not the case **with 3 failed test** when I removed the `nil`
I have been caught out when I put a puts "whatever" and it happens to be the last thing in the method. I think the same thing is going on the last thing is the errors  array

Reply:
Well thanks for you comments, it got me thinking so this works, which I think is better

```
  def x509_certificate
    if @last_converted_value || value != @last_converted_value
      @last_converted_value = value
      convert_value_to_x509_certificate
    end
  rescue
    errors.add(:certificate, 'is not a valid x509 certificate')
    nil
  end
```

Reply:
yes, that's what I meant... also, if you remove the if condition it's still all good, like:

```
def x509_certificate
    convert_value_to_x509_certificate
  rescue
    errors.add(:certificate, 'is not a valid x509 certificate') if @x509.nil?
    nil
  end
```

Reply:
I think this is less efficient though b'cos 
```
begin
      OpenSSL::X509::Certificate.new(value)
    rescue
      OpenSSL::X509::Certificate.new(Base64.decode64(value))
    end
```
would be called every time there is a reference to **x509_certificate** that's potential five times for the upload certificate event
as opposed to just **once**  if the if were there.

Reply:
So made the change as requested also read somewhere to favour readability over optimisation so this may just be one of those cases.

Reply:
in that case we should extract the x509_certificate as an instance variable at the top and use it around instead of doing this "hard-to-read" if statement

Reply:
I can show you what I mean on Monday. Have a good weekend!